pkg/driver/controller.go

```diff
@@ -485,7 +486,84 @@ func (d *Driver) ListSnapshots(ctx context.Context, req *csi.ListSnapshotsReques
 }
 
 func (d *Driver) ControllerExpandVolume(ctx context.Context, req *csi.ControllerExpandVolumeRequest) (*csi.ControllerExpandVolumeResponse, error) {
-	return nil, status.Error(codes.Unimplemented, "")
+	klog.V(4).Infof("ControllerExpandVolume: called with args %+v", *req)
+	var err error
+
+	volumeID := req.GetVolumeId()
+	if len(volumeID) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "Volume ID not provided")
+	}
+
+	capRange := req.GetCapacityRange()
+	if capRange == nil {
+		return nil, status.Error(codes.InvalidArgument, "Capacity range not provided")
+	}
+
+	if capRange.GetLimitBytes() > 0 && capRange.GetRequiredBytes() > capRange.GetLimitBytes() {
+		return nil, status.Errorf(codes.OutOfRange, "Requested storage capacity of %d bytes exceeds capacity limit of %d bytes.", capRange.GetRequiredBytes(), capRange.GetLimitBytes())
+	}
+
+	newCapacity := util.BytesToGiB(capRange.GetRequiredBytes())
+
+	var currentCapacity int64
+	splitVolumeId := strings.SplitN(volumeID, "-", 2)
+	if splitVolumeId[0] == cloud.FilesystemPrefix {
+		var fs *cloud.FileSystem
+		fs, err = d.cloud.DescribeFileSystem(ctx, volumeID)
+		if fs != nil {
+			currentCapacity = fs.StorageCapacity
+		}
+	} else if splitVolumeId[0] == cloud.VolumePrefix {
+		var v *cloud.Volume
+		v, err = d.cloud.DescribeVolume(ctx, volumeID)
+		if v != nil {
+			currentCapacity = v.StorageCapacityQuotaGiB
+		}
+	} else {
+		err = cloud.ErrNotFound
+	}
+
+	if err != nil {
+		if err == cloud.ErrNotFound {
+			return nil, status.Error(codes.NotFound, "Volume not found")
+		}
+		return nil, status.Errorf(codes.Internal, "Could not get volume with ID %q: %v", volumeID, err)
+	}
+
+	if newCapacity <= currentCapacity {
+		// Current capacity is sufficient to satisfy the request
+		klog.V(4).Infof("ControllerExpandVolume: current filesystem capacity of %d GiB matches or exceeds requested storage capacity of %d GiB, returning with success", currentCapacity, newCapacity)
+		return &csi.ControllerExpandVolumeResponse{
+			CapacityBytes:         util.GiBToBytes(currentCapacity),
+			NodeExpansionRequired: false,
+		}, nil
+	}
+
+	var finalCapacity *int64
+	if splitVolumeId[0] == cloud.FilesystemPrefix {
+		finalCapacity, err = d.cloud.ResizeFileSystem(ctx, volumeID, newCapacity)
+	} else if splitVolumeId[0] == cloud.VolumePrefix {
+		finalCapacity, err = d.cloud.ResizeVolume(ctx, volumeID, newCapacity)
+	}
+
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "resize failed: %v", err)
+	}
+
+	if splitVolumeId[0] == cloud.FilesystemPrefix {
+		err = d.cloud.WaitForFileSystemResize(ctx, volumeID, *finalCapacity)
+	} else if splitVolumeId[0] == cloud.VolumePrefix {
+		err = d.cloud.WaitForVolumeResize(ctx, volumeID, *finalCapacity)
+	}
+
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "filesystem is not resized: %v", err)
```

Comment:
Same comment as above: it looks like this error message could apply to volumes and file systems, yet you specifically state "filesystem" here. I'd also put the file system/volume id in this error message.